controllers/apidefinition_controller.go

```diff
@@ -132,29 +132,54 @@ func (r *ApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		// we support only one certificate secret name for mvp
 		if len(desired.Spec.CertificateSecretNames) != 0 {
 			certName := desired.Spec.CertificateSecretNames[0]
-			tykCertID, err := r.checkSecretAndUpload(ctx, certName, namespacedName, log, &env)
+			tykCertID, err := r.checkSecretAndUpload(ctx, certName, namespacedName.Namespace, log, &env, "")
 			if err != nil {
 				return err
 			}
 
 			upstreamRequestStruct.Spec.Certificates = []string{tykCertID}
 		}
 
+		// Check Pinned Public keys
+		if len(desired.Spec.PinnedPublicKeysSecretNames) != 0 {
+			if upstreamRequestStruct.Spec.PinnedPublicKeys == nil {
+				upstreamRequestStruct.Spec.PinnedPublicKeys = map[string]string{}
+			}
+
+			for domain, secretName := range desired.Spec.PinnedPublicKeysSecretNames {
+				// Set the namespace for referenced secret to the current namespace where ApiDefinition lives.
+				tykCertID, err := r.checkSecretAndUpload(ctx, secretName, req.Namespace, log, &env, "public-key")
+				if err != nil {
+					// we should log the missing secret, but we should still create the API definition
+					log.Info(
```

Comment:
Shouldn't it be an error?

Reply:
If users target missing secret, we are creating the ApiDefinition as it is except Public Key Pinning feature. Once the secret is created, secret_cert controller creates the certificate and update the ApiDefinition's Public Key Pinning field. This is the same flow in the Upstream Certificate https://github.com/TykTechnologies/tyk-operator/blob/master/controllers/apidefinition_controller.go#L116